libtsuba/include/tsuba/Cache.h

```diff
@@ -0,0 +1,182 @@
+// parallel-hashmap is a lock ordering problem.  parallel-hashmap 1.33 allows
+// execution of code with the write lock held, only when modifying, not when adding an
+// element.  We have to modify the LRU list when eviciting an element, so the natural
+// lock ordering is parallel-hashmap write lock, then list lock.  But without a way to
+// execute insert code with the parallel-hashmap write lock held, it seemed like there
+// would be some form of race condition.
+
+#include <list>
+#include <optional>
+#include <string>
+
+#include <boost/container_hash/hash.hpp>
+
+#include "katana/ArrowInterchange.h"
+#include "katana/ConcurrentMap.h"
+#include "katana/Logging.h"
+
+namespace tsuba {
+
+struct KATANA_EXPORT KeyBase {
+  enum class NodeEdge { kNode, kEdge };
+  bool operator==(const KeyBase& o) const {
+    return node_edge_ == o.node_edge_ && name_ == o.name_;
+  }
+
+  NodeEdge node_edge_;
+  std::string name_;
+
+  struct Hash {
+    std::size_t operator()(const KeyBase& key) const {
+      using boost::hash_combine;
+      using boost::hash_value;
+
+      std::size_t seed = 0;
+      hash_combine(seed, hash_value(key.node_edge_));
+      hash_combine(seed, hash_value(key.name_));
+
+      // Return the result.
+      return seed;
+    }
+  };
+};
+
+template <typename Value>
+class KATANA_EXPORT Cache {
+public:
+  using ListType = std::list<KeyBase>;
+  struct MapValue {
+    Value value;
+    // This allows us to delete the old position in the LRU list without a scan
+    typename ListType::iterator lruit;
+  };
+  using MapType = std::unordered_map<KeyBase, MapValue, KeyBase::Hash>;
+  enum class ReplacementPolicy { kLRUSize, kLRUBytes };
+
+  Cache(
+      ReplacementPolicy policy = ReplacementPolicy::kLRUSize,
+      size_t lru_capacity = 0, size_t bytes_capacity = 0,
+      std::function<void(const KeyBase& key)> evict_cb = nullptr,
+      std::function<size_t(const Value& value)> value_to_bytes = nullptr)
+      : policy_(policy),
+        lru_capacity_(lru_capacity),
+        bytes_capacity_(bytes_capacity),
+        evict_cb_(std::move(evict_cb)),
+        value_to_bytes_(std::move(value_to_bytes)) {
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUSize || lru_capacity_ > 0,
+        "kLRUSize policy requires non-zero LRU entries");
+    KATANA_LOG_VASSERT(
+        policy_ != ReplacementPolicy::kLRUBytes || bytes_capacity_ > 0,
+        "kLRUBytes policy requires non-zero byte capacity");
+  }
+
+  size_t size() const { return key_to_value_.size(); }
+  size_t bytes() const { return total_bytes_; }
+
+  size_t LRUSizeCapacity() const { return lru_capacity_; }
+  size_t LRUBytesCapacity() const { return bytes_capacity_; }
+
+  bool Empty() const { return key_to_value_.empty(); }
+
+  bool Contains(const KeyBase& key) {
+    return key_to_value_.find(key) != key_to_value_.end();
+  }
+
+  void Insert(const KeyBase& key, const Value& value) {
```

Comment:
Why not just template on the key type as well? and thus support any hashable thing?

Reply:
What I wanted to say is, "your key needs to be able to hash itself."  I thought a pure template would not do that, so I provided a base class for any key class.  It seemed kind of like a C++-20 concept, but I'm not sure of that either.

What is the best way to say that I need to know if your key is for node or edge, it should have a name (that will be unique for nodes and unique for edges) and I need to know how to hash it.